cli/azd/design-notes/environments.md

```diff
@@ -0,0 +1,271 @@
+
+The deployment parameters are stored in `config.json` under the `infra` section, segmented by module.
+
+```json
+{
+  "id": "9fdfa455-2131-4faf-b2e3-8380aba97c96",
+  "name": "default",
+  "infra": {
+    "main": { 
+      "properties": {
+         // ... the object that is set as the value of the "properties" key of a deploymentParameters file
+      }
+    }
+  }
+}
+```
+
+To save ourselves the trouble of designing our own format, we reuse the schema of a `deploymentParamters` file. One nice feature here is that system provides enough flexibility to express references to secrets in Azure Key Vault, which will be a useful feature for us down the line.
+
+We create a new set of `azd infra config` commands with actions like `set`, `get`, `list`, `delete` to allow management of these parameters without having to edit the `config.json` file.
+
+### Improved Deployment Configuration
+
+Since we now operate directly on deployment template parameters, we can leverage existing template metadata (like the description of a parameter) to provide template specific help on how the parameter is used as well as validation.
+
+For example, our ToDo application could be updated to have the following "name" parameter:
+
+```bicep
+@minLength(1)
+@maxLength(64)
+@description('The name used as a prefix for the resource group which holds all the resources for the ToDo application. The suffix `-rg` will automatically be added.')
+param name string
+```
+
+Which provides template specific information on how the parameter is used. It also ensures that these templates behave nicely with other deployment tools (like creating a deployment from the portal or using `az` directly).
+
+We leverage the existing metadata system for bicep templates and parameters to allow some `azd` specific metadata to be applied to properties:
+
+```bicep
+@description('The Azure location to use for all resources.')
+@metadata({
+    azd: {
+        type: 'location'
+    }
+})
+param location string
+```
+
+The `type` property of the `azd` metadata object denotes that this value should be set to an Azure location name (e.g. "westus2") and is used by `azd` to offer a "pick a location from this list of all Azure locations). This composes with other validation decorators that you can apply, for example `@allowed()` which can be used to restrict the set of locations offered to the user to pick from (for cases where a template uses services that are not deployed to all regions):
+
+```bicep
+@description('The Azure location to use for all resources.')
+@allowed([
+    'westus2'
+    'centralus'
+])
+@metadata({
+    azd: {
+        type: 'location'
+    }
+})
```

Comment:
I'm not familiar with this `metadata` field and we would need to ensure a similar concept is available on other provider implementations.  We may just need to fall back and use a naming convention that we look for for parameters across different provision providers.

Reply:
> We may just need to fall back and use a naming convention that we look for for parameters across different provision providers.

Alternately you could imagine encoding this metadata in JSON and storing it in an azd specific file side by side with whatever module defined the parameters and `azd` could use that file (if it exists).